services/brig/test/integration/API/User/Auth.hs

```diff
@@ -456,6 +457,52 @@ testLoginVerify6DigitResendCodeSuccessAndRateLimiting brig galley _opts db = do
         (Just defCookieLabel)
         (Just $ Code.codeValue mostRecentCode)
 
+testLoginVerify6DigitLimitRetries :: Brig -> Galley -> Opts.Opts -> DB.ClientState -> Http ()
+testLoginVerify6DigitLimitRetries brig galley _opts db = do
+  (u, tid) <- createUserWithTeam' brig
+  let Just email = userEmail u
+  let checkLoginFails body =
+        login brig body PersistentCookie !!! do
+          const 403 === statusCode
+          const (Just "code-authentication-failed") === errorLabel
+
+  Util.setTeamFeatureLockStatus @Public.SndFactorPasswordChallengeConfig galley tid Public.LockStatusUnlocked
+  Util.setTeamSndFactorPasswordChallenge galley tid Public.FeatureStatusEnabled
+  Util.generateVerificationCode brig (Public.SendVerificationCode Public.Login email)
+  key <- Code.mkKey (Code.ForEmail email)
+  Just correctCode <- Util.lookupCode db key Code.AccountLogin
+  let wrongCode = Code.Value $ unsafeRange (fromRight undefined (validate "123456"))
+  -- login with wrong code should fail 3 times
```

Comment:
The three expressions look equal: How about using a loop expression? E.g. [`forM_ [1..3] ...`](https://hackage.haskell.org/package/base-4.17.0.0/docs/Control-Monad.html#v:forM_)